### PR TITLE
Add .indexInBrackets array serialization option for URL parameters

### PR DESCRIPTION
**Pull request by TiagoMaiaL**
### Issue Link :link:
- #1839 
- #2431

### Goals :soccer:

There is no real specification for how to encode URL parameters like arrays and dictionaries.

For `["foo": ["a", 1, true, ["bar": 2], ["qux": 3], ["quy": ["quz": 3]]]]`

1. default `.brackets` encode is:
`foo[]=a&foo[]=1&foo[]=1&foo[][bar]=2&foo[][qux]=3&foo[][quy][quz]=3`

2. `.noBrackets` encode is:
`foo=a&foo=1&foo=1&foo[bar]=2&foo[qux]=3&foo[quy][quz]=3`

3. [jQuery](https://api.jquery.com/jQuery.param/#example-1) >=1.4 and Node.js package [query-string](https://www.npmjs.com/package/query-string) provide index style:
`foo[0]=a&foo[1]=1&foo[2]=1&foo[3][bar]=2&foo[4][qux]=3&foo[5][quy][quz]=3`

### Implementation Details :construction:

A new `ArrayEncoding` enumeration case is added to both `URLEncoding` and `URLEncodedFormEncoder`. The cases now are `.brackets` and `.noBrackets` and `.indexInBrackets`. The `arrayEncoding` option passed to the initializer is used to control the array key serialization in `URLEncodedFormEncoder` and `URLEncoding`.

### Testing Details :mag:

New tests were added to the `ParameterEncoderTests` and `ParameterEncodingTests` suites. They follow the same naming convention but with an  appended `WithIndexInBrackets` suffix.

### Note:

This PR continues the work started at #3015.

### CODE REVIEW

File: Source/ParameterEncoding.swift [hunk 1]
```diff
@@ -85,13 +85,17 @@ public struct URLEncoding: ParameterEncoding {
         case brackets
         /// No brackets are appended. The key is encoded as is.
         case noBrackets
+        /// Brackets containing the item index appended. This matches the jQuery and Node.js behavior.
```
Comment (jshier) [inline Source/ParameterEncoding.swift:88]:
```suggestion
        /// Brackets containing the item index are appended. This matches the jQuery and Node.js behavior.
```

File: Source/ParameterEncoding.swift [hunk 2]
```diff
@@ -85,13 +85,17 @@ public struct URLEncoding: ParameterEncoding {
...
+        case indexInBrackets
 
-        func encode(key: String) -> String {
+        func encode(key: String, index: Int) -> String {
```
Comment (jshier) [inline Source/ParameterEncoding.swift:91]: I'd suggest something like `encode(key: String, atIndex index: Int)` to better match Swift's naming guidelines to make the usage a bit more fluent.

File: Source/URLEncodedFormEncoder.swift [hunk 3]
```diff
@@ -48,15 +48,19 @@ public final class URLEncodedFormEncoder {
...
         case noBrackets
+        /// Brackets containing the item index appended. This matches the jQuery and Node.js behavior.
+        case indexInBrackets
 
         /// Encodes the key according to the encoding.
         ///
-        /// - Parameter key: The `key` to encode.
-        /// - Returns:       The encoded key.
-        func encode(_ key: String) -> String {
+        /// - Parameter key:    The `key` to encode.
```
Comment (jshier) [inline Source/URLEncodedFormEncoder.swift:56]: Now that there are multiple parameters, you can regenerate these docs using `Command - Option - /` and get a `Parameters` section instead. Also, though it was missing before, our general style is to put an empty line between the parameters and the return.

File: Source/URLEncodedFormEncoder.swift [hunk 4]
```diff
@@ -48,15 +48,19 @@ public final class URLEncodedFormEncoder {
...
+        /// - Parameter index:  When this enum instance is `.indexInBrackets`, the `index` to encode.
+        /// - Returns:          The encoded key.
+        func encode(_ key: String, index: Int) -> String {
```
Comment (jshier) [inline Source/URLEncodedFormEncoder.swift:59]: Like the other `encode`, `index` would be more fluently expressed as `atIndex`.

**Review by jshier — CHANGES_REQUESTED** [review]
Thanks for the PR, this looks great! Just a few doc formatting suggestions and some naming and this should be all set.

File: Source/URLEncodedFormEncoder.swift [hunk 5]
```diff
@@ -930,8 +936,8 @@ final class URLEncodedFormSerializer {
     }
 
     func serialize(_ array: [URLEncodedFormComponent], forKey key: String) -> String {
-        var segments: [String] = array.map { component in
-            let keyPath = arrayEncoding.encode(key)
+        var segments: [String] = array.enumerated().map { index, component in
+            let keyPath = arrayEncoding.encode(key, index: index)
```
Comment (jshier) [inline Source/URLEncodedFormEncoder.swift:940]: Seems you've missed the use site for the updated name.

**Review by jshier — CHANGES_REQUESTED** [review]
I think you missed one use site of the updated naming. Once that's updated I can trigger the test suite.

File: Source/URLEncodedFormEncoder.swift (hunk 5 above)
Comment (TiagoMaiaL) [reply Source/URLEncodedFormEncoder.swift]: Thanks for catching it, I forgot to build  🤦🏽‍♂️.

**Review by jshier — APPROVED** [review]
👍

**Comment by kleverson** [comment]
Great work!

**Pull request merged**